Fix rtl930x VLAN, LEDs and status

0xharshal opened the pull request:
- Fix VLAN behaviour
- support jumbo frames by default
- enable 2.5 gig for rtl930x
- Fix led behaviour
- Fix detection of current port status
- Expose register and phy_mmd register r/w in sysfs
  to aid development and further debugging

0xharshal committed:
rtl838x: Enable jumbo frames by default

Increase DEFAULT_MTU and max-mtu size
Increase truncate length on rx of jumbo frame

Acked-by: Simon Wunderlich <sw@simonwunderlich.de>
Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>

0xharshal committed:
rtl83xx: dsa: Fix VLAN filtering for cpu port and non-cpu ports

Before driver code
 - enabled egress filter for cpu and non-cpu ports
 - enabled ingress filter for non-cpu ports

This patch explicitly enables ingress and egress filtering for
non-cpu ports and disables ingress and egress filtering for cpu
port.

Acked-by: Simon Wunderlich <sw@simonwunderlich.de>
Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>

0xharshal committed:
rtl83xx: dsa: Do nothing when vid 0

Following other dsa drivers, vid 0 is no-op

Acked-by: Simon Wunderlich <sw@simonwunderlich.de>
Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>

0xharshal committed:
rtl930x: Rework per port LED configuration

Use led_setX to determine number of LEDs per port
Introduce macros to calculate register value and shift
for particular LED in a particular set

Problem with previous implementation is that it uses is10G status to
determine leds per port. However with usxgmii, driver sets 10g, 5g and 2.5g
so even though there are only 2 leds per port it selects 4 leds per port

This implementation relies on configured led_set node

Acked-by: Simon Wunderlich <sw@simonwunderlich.de>
Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>

0xharshal committed:
rtl83xx: dsa: Clear duplex bit correctly

Without this, luci shows 10M full duplex when there is no link
So explicitly set half duplex and unknown speed

Acked-by: Simon Wunderlich <sw@simonwunderlich.de>
Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>

mr-wewlad commented:
Hello,

Just wanted to note that the December 17 snapshot build for the ZyXEL XGS1250-12 seems to have issues with layer 2 switching. After some help/pointers from @paulfertser I eventually ran into this PR and manually compiled it.

After flashing this OpenWrt build most functionality seems to be working as expected! :) Mind you I'm currently just using it as a simple managed layer 2 switch, not doing anything fancy.

FWIW, in my experience this PR significantly improves the functionality of OpenWRT on the ZyXEL XGS1250-12.

Borromini commented:
@svanheule Would you mind having a look?

TobleMiner commented:
Thanks for this great contribution! At first glance this MR really seems to fix the QinQ issues I was also seeing previously.

I have only one small comment: On the ZyXEL XGS1250-12 the LEDs are no longer correctly mapped with the LED patches applied.
I feel like either:
1. The `led_set` dt node for the ZyXEL XGS1250-12 was never correct (or semantics of `led_set` change with this PR)
2.  [This line](https://github.com/openwrt/openwrt/pull/14221/files#diff-e1daddd69ff468fcba6b334777267b3219ac1facd8d77773f27de9f8a4e4f822R2410) might be off by one and should rather be `led_sets = i + 1;`

Unfortunately I'm not really an expert for rtl93xx so please excuse if I'm ignorant of something that should be obvious here :smile:

svanheule commented on target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c:
```diff
@@ -2396,23 +2397,34 @@ static void rtl930x_led_init(struct rtl838x_switch_priv *priv)
+		v = be32_to_cpup(led_set) << 16 | be32_to_cpup(led_set + 1);
+		sw_w32(v, RTL930X_LED_SET0_0_CTRL - 4 - i * 8);
+		v = be32_to_cpup(led_set + 2) << 16 | be32_to_cpup(led_set + 3);
+		sw_w32(v, RTL930X_LED_SET0_0_CTRL - i * 8);
```
Some mental notes:
Index nr 3 is written to the lowest bits of `SETi_0_CTRL`, index nr 2 to the highest bits, index nr 1 to the lowest of `SETi_1_CTRL`, index nr 0 the hightest bits. At first glance, this means that structure is `led_setX = <LED3 LED2 LED1 LED0>` (which is a bit backwards, but OK).

You might be able use the `LEDx` values to determine how many LEDs a port is configured to use, but I see that for the XGS1250-10 the `LED2` value is set to 0xffff, i.e. matching all speeds and other conditions, which is also strange.

0xharshal replied:
LED setup reworked, check latest commits

svanheule commented on target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/debugfs.c:
```diff
@@ -726,4 +829,7 @@ void rtl930x_dbgfs_init(struct rtl838x_switch_priv *priv)
 	debugfs_create_file("l2_table", 0400, dbg_dir, priv, &l2_table_fops);
+
+	debugfs_create_file("reg_rw", 0x200, dbg_dir, priv, &reg_rw_fops);
+	debugfs_create_file("mmd_rw", 0x200, dbg_dir, priv, &mmd_rw_fops);
```
Given that `/dev/mem` exists for register access, and [`phytool`](https://github.com/wkz/phytool) could be (packaged and) used for MMD access, what makes this worthwhile to merge? You're free to keep this patch around, but I'm reluctant to merge stuff like this if there are equivalent tools available.

0xharshal replied:
dropped :+1:

svanheule reviewed (changes requested):
Thanks for the patches! I think you'll need to have another look at the `led_set` patch though.

Did you co-develop these changes with Simon? If so, then you should also add [`Co-developed-by:` tags](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/submitting-patches.rst?id=HEAD#n491)

svanheule commented:
@0xharshal @TobleMiner If you would like access to edit info at https://svanheule.net/realtek, I can add  your Github accounts to the whitelist.

0xharshal commented on target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c:
```diff
@@ -135,10 +135,6 @@ static void rtl83xx_vlan_setup(struct rtl838x_switch_priv *priv)
-
-	/* Set forwarding action based on inner VLAN tag */
-	for (int i = 0; i < priv->cpu_port; i++)
-		priv->r->vlan_fwd_on_inner(i, true);
```
Consider following VLAN configuration,

VLAN_ID | lan0 | lan1 | lan2 | lan3 | lan4 | lan5 | lan6 | lan7
--------|------|------|------|------|------|------|------|------
10      |  -   |    U |    U |      |      |      |      |      
20      |  -   |    - |    - |    U |    U |      |      |  T
30      |  -   |    - |    - |      |      |   U  |    U |  U

send Untagged packet in **lan1**
- expected: untagged packet on **lan2**
- w/o patch: receives tagged (with id:10) on **lan2**

send tagged packet(vid:X, where X!=10) in **lan1** 
- expected: nothing should be received on **lan2**
- w/o patch: receives tagged packet without modification on **lan2**

send tagged packet(vid:10) in **lan1**
- expected: not sure! what the desired behavior is in this case
- w/o patch: receives tagged packet without modification on **lan2**

send untagged packet in **lan3**
- expected: packet should be received with added tag vid:20 on **lan7**
- w/o patch: receives QinQ with [vid:20[vid:20]]

send tagged packet(vid:X, where X!=20) in **lan3**
- expected: packet should be dropped
- w/o patch: receives QinQ with [vid:20[vid:X]] on **lan7**

send tagged packet(vid:20) in **lan3**
- expected: not sure!
- w/o patch: receives QinQ with [vid:20[vid:20]] on **lan7**

send tagged packet(vid:20) in **lan7**
- expected: packet should be received w/o tag on **lan4**
- w/o patch: packet is dropped

I need some help on summarizing this behavior, how about, 
```
Fix 802.1QinQ for trunk ports

Before, packets would contain two tags for trunk port 
and packets would stay tagged for access ports.
```

svanheule replied:
Maybe something like:

"Before, outgoing packets would receive an extra tag, corresponding to the PVID of the port. On untagged ports, this meant outgoing packets with a single tag. On tagged ports, this meant outgoing QinQ packets, where the inner tag was either the PVID of the untagged ingress port, or the already assigned original (single) tag."

I'm not sure what the behavior should be for tagged frames on untagged ports. Might even be configurable (drop or strip).

svanheule replied:
And thanks for investigating!

0xharshal replied:
Made the changes explicit, use `vlan_fwd_on_inner(i, false)`

0xharshal commented on target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c:
```diff
@@ -136,14 +136,6 @@ static void rtl83xx_vlan_setup(struct rtl838x_switch_priv *priv)
-		priv->r->vlan_port_pvid_set(i, PBVLAN_TYPE_OUTER, 0);
-		priv->r->vlan_port_pvidmode_set(i, PBVLAN_TYPE_INNER, PBVLAN_MODE_UNTAG_AND_PRITAG);
-		priv->r->vlan_port_pvidmode_set(i, PBVLAN_TYPE_OUTER, PBVLAN_MODE_UNTAG_AND_PRITAG);
-	}
```
Yes I think the driver is setting pvid to 0
OpenWRT shows PVID 1 with `bridge vlan` command


And considering rtlsdk, they also set pvid=1 by default, 
https://gitlab.com/olliver/openwrt/realtek_sdk/-/blob/xgs1250/sdk/include/rtk/default.h?ref_type=heads#L46

0xharshal replied:
Some more investigation before this commit,
on the first boot, following network config is generated,
```
config interface 'loopback'                                                                                                                                                                                                                                   
        ...                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                              
config globals 'globals'                                                                                                                                                                                                                                      
        ...                                                                                                                                                                                                                            
                                                                                                                                                                                                                                                              
config device 'switch'                                                                                                                                                                                                                                        
        option name 'switch'                                                                                                                                                                                                                                  
        option type 'bridge'                                                                                                                                                                                                                                  
        option macaddr '...'                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                              
config bridge-vlan 'lan_vlan'                                                                                                                                                                                                                                 
        option device 'switch'                                                                                                                                                                                                                                
        option vlan '1'                                                                                                                                                                                                                                       
        option ports 'lan0 lan1 lan2 lan3 lan4 lan5 lan6 lan7'                                                                                                                                                                                                
                                                                                                                                                                                                                                                              
config device                                                                                                                                                                                                                                                 
        option name 'switch.1'                                                                                                                                                                                                                                
        option macaddr '...'                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                              
config interface 'lan'                                                                                                                                                                                                                                        
        option device 'switch.1'                                                                                                                                                                                                                              
        option proto 'static'                                                                                                                                                                                                                                 
        option ipaddr '192.168.1.1'                                                                                                                                                                                                                           
        option netmask '255.255.255.0'                                                                                                                                                                                                                        
        option ip6assign '60'                                                                                                                                                                                                                                 
                                                    
```
In this configuration `bridge vlan` shows following output,
```
root@OpenWrt:/# bridge vlan 
port              vlan-id  
lan0              1 PVID Egress Untagged
lan1              1 PVID Egress Untagged
lan2              1 PVID Egress Untagged
lan3              1 PVID Egress Untagged
lan4              1 PVID Egress Untagged
lan5              1 PVID Egress Untagged
lan6              1 PVID Egress Untagged
lan7              1 PVID Egress Untagged
switch            1
```
And in this state, I am not able to ping a device connected to one of the lan ports from switch, `ping 192.168.1.100` no packets are received on any of the lan ports

---

If I make following changes to the `/etc/config/network`,
```
--- a/etc/config/network      2024-01-03 13:30:20.130108139 +0100
+++ b/etc/config/network      2024-01-03 13:30:36.321918906 +0100
@@ -10,19 +10,11 @@
 config device 'switch'
         option name 'switch'
         option type 'bridge'
-        option macaddr '...'
-
-config bridge-vlan 'lan_vlan'
-        option device 'switch'
-        option vlan '1'
         option ports 'lan0 lan1 lan2 lan3 lan4 lan5 lan6 lan7'
-
-config device
-        option name 'switch.1'
         option macaddr '...'
 
 config interface 'lan'
-        option device 'switch.1'
+        option device 'switch'
         option proto 'static'
         option ipaddr '192.168.1.1'
         option netmask '255.255.255.0'

```
output of `bridge vlan`,
```
root@OpenWrt:/# bridge vlan                                                                                                                                                                                                                                   
port              vlan-id                                                                                                                                                                                                                                     
lan0              1 PVID Egress Untagged                                                                                                                                                                                                                      
lan1              1 PVID Egress Untagged                                                                                                                                                                                                                      
lan2              1 PVID Egress Untagged                                                                                                                                                                                                                      
lan3              1 PVID Egress Untagged                                                                                                                                                                                                                      
lan4              1 PVID Egress Untagged                                                                                                                                                                                                                      
lan5              1 PVID Egress Untagged                                                                                                                                                                                                                      
lan6              1 PVID Egress Untagged                                                                                                                                                                                                                      
lan7              1 PVID Egress Untagged                                                                                                                                                                                                                      
switch            1 PVID Egress Untagged          
```
Notice that switch vlan-id 1 is now PVID Egress Untagged.
After this, I am able to ping the device connected to one of the lan ports.

---
If I remove reset-pvid for-loop I am able to ping the connected device just fine without making any changes. and here I am not sure if the default config is problematic or the driver code. 
Also it looks like driver is setting pvid 0, but default config comes up with pvid 1.

0xharshal replied:
Well I just spotted the for-loop goes over 28 ports, and that was the issue.
- reset pvid is anyways overwritten by the configuration so this is misinterpretation on my side of setup code.
```
diff --git a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c
index 9b26231a71..83bf92c8d7 100644
--- a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c
+++ b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c
@@ -137,7 +137,7 @@ static void rtl83xx_vlan_setup(struct rtl838x_switch_priv *priv)
                priv->r->vlan_set_tagged(i, &info);
 
        /* reset PVIDs; defaults to 1 on reset */
-       for (int i = 0; i <= priv->ds->num_ports; i++) {
+       for (int i = 0; i < priv->cpu_port; i++) {
                priv->r->vlan_port_pvid_set(i, PBVLAN_TYPE_INNER, 0);
                priv->r->vlan_port_pvid_set(i, PBVLAN_TYPE_OUTER, 0);
                priv->r->vlan_port_pvidmode_set(i, PBVLAN_TYPE_INNER, PBVLAN_MODE_UNTAG_AND_PRITAG);
```

before the code goes up to i = 29, which calculates the register value to `0x834c` which is wrong, but still not the root cause since that register is being overwritten by vlan_fwd_on_inner.
Problems is when register 0x8348 (VLAN_PORT_PB_VLAN[28]) is set with 0x000000 with
```
priv->r->vlan_port_pvid_set(i:28, PBVLAN_TYPE_INNER, 0);
priv->r->vlan_port_pvid_set(i:28, PBVLAN_TYPE_OUTER, 0);
priv->r->vlan_port_pvidmode_set(i:28, PBVLAN_TYPE_INNER, PBVLAN_MODE_UNTAG_AND_PRITAG);
priv->r->vlan_port_pvidmode_set(i:28, PBVLAN_TYPE_OUTER, PBVLAN_MODE_UNTAG_AND_PRITAG);
```
If I manually set register `0x834c` to `0x00010004` (OPVID:1,IPVID:1) ping starts working. bootloader and rtlsdk sets (OPVID:1, IPVID:1) so what do you suggest the code should do? set default pvid to 1 for all ports or do this just for cpu port?

0xharshal replied:
use reset pvid 1 instead of 0, check latest commit's message and above investigation for explanation.

0xharshal commented on target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c:
```diff
@@ -2396,23 +2397,34 @@ static void rtl930x_led_init(struct rtl838x_switch_priv *priv)
-		if (priv->ports[i].phy_is_integrated)
-			v = 0x1;
-		sw_w32_mask(0x3 << pos, v << pos, RTL930X_LED_PORT_NUM_CTRL(i));
+		sw_w32_mask(0x3 << pos, led_sets << pos, RTL930X_LED_PORT_NUM_CTRL(i));
```
Okay I checked again, and I think the DTS, my code and old code all are wrong,
See, https://gitlab.com/olliver/openwrt/realtek_sdk/-/blob/xgs1250/sdk/system/include/hwp/hw_profiles/xgs1250_12.c?ref_type=heads#L36

There are 3 LED sets for xgs1250,
led_set0 -> has 2 LEDs and 0xffff because of `HWP_LED_END`
led_set1 -> has 4 LEDs
led_set2 -> has 2 LEDs and 0xffff because of `HWP_LED_END`

0xffff can also be read as enable LED to indicate everything

Looking at Datasheet of xgs1250-12, https://www.zyxelguard.com/datasheets/XGS1250-12_1.pdf
- for 1G ports, 2 leds per port 1-8. 
  - Link/Act + 10/100 M
  - 1000 M
- for 10G ports, 5 color LEDs per port 9,10,11 - which is achieved by combining 4 color LEDs 
  - 100M
  - 1000M
  - 2.5G
  - 5G
  - 10G
- for 10G sfp port
  - 1000M
  - 10G

I don't have xgs1250 to see how it works, i will rework LED code and test on a dev board that I have and back, after someone can compile and check on their xgs1250

0xharshal replied:
reworked led code, check latest push

0xharshal committed:
rtl93xx: dsa: Fix 802.1QinQ for trunk ports

Fix incorrect register value being set for VLAN_PORT_FWD

Before, the 0b1111 would be set for the register which means
outgoing packets would receive an extra tag, corresponding
to the PVID of the port.

On untagged ports, this meant outgoing packets
with a single tag.

On tagged ports, this meant outgoing QinQ packets,
where the inner tag was either the PVID of the
untagged ingress port, or the already assigned
original (single) tag.

Acked-by: Simon Wunderlich <sw@simonwunderlich.de>
Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>

0xharshal committed:
rtl83xx: dsa: reset PVID to 1 instead of 0

Before, PVID is reset for all ports and goes out of bounds.
Also, PVID is later changed by dsa configuration by `ip link`
and `bridge vlan` commands, this does not change the CPU port pvid
and CPU pvid stays 0. It does not allow sending packets from OpenWRT
to any connected devices unless default configuration is changed

This change iterates upto and including cpu_port and sets default PVID
to 1.
for lan* ports pvid can be configured with `ip link` and
`bridge vlan` commands

Acked-by: Simon Wunderlich <sw@simonwunderlich.de>
Signed-off-by: Harshal Gohel <hg@simonwunderlich.de>

0xharshal commented on target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c:
```diff
@@ -249,7 +249,7 @@ static void rtl930x_vlan_fwd_on_inner(int port, bool is_set)
 	/* Always set all tag modes to fwd based on either inner or outer tag */
 	if (is_set)
-		sw_w32_mask(0, 0xf, RTL930X_VLAN_PORT_FWD + (port << 2));
+		sw_w32_mask(0xf, 0xf, RTL930X_VLAN_PORT_FWD + (port << 2));
```
dropped

0xharshal commented on target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c:
```diff
@@ -806,7 +806,7 @@ static void rtl93xx_hw_en_rxtx(struct rtl838x_eth_priv *priv)
 
 	if (priv->family_id == RTL9300_FAMILY_ID)
-		sw_w32(0x217, priv->r->mac_force_mode_ctrl + priv->cpu_port * 4);
+		sw_w32(0x22F, priv->r->mac_force_mode_ctrl + priv->cpu_port * 4);
```
dropped the change, it made no difference for now, will reintroduce if required

0xharshal commented on target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c:
```diff
@@ -918,6 +918,8 @@ static int rtl838x_eth_open(struct net_device *ndev)
 		/* Trap MLD and IGMP messages to CPU_PORT */
 		sw_w32((0x2 << 3) | 0x2,  RTL930X_VLAN_APP_PKT_CTRL);
+		/* Disable truncate jumbo rx frame */
+		sw_w32_mask(BIT(6), 0, RTL930X_DMA_IF_CTRL);
```
introduced default mtu for rtl838x, rtl839x and rtl93xx, please verify in latest push

svanheule reviewed (commented):
Please make sure all your commit messages have a Signed-off-by tag.

Please also clarify what Simon's involvement is in these patches, or I'm going to have to ask you to drop his Signed-off-by.

0xharshal commented on target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c:
```diff
@@ -146,7 +146,7 @@ static void rtl83xx_vlan_setup(struct rtl838x_switch_priv *priv)
 
 	/* Set forwarding action based on inner VLAN tag */
```
I think comment now is correct, but function args and function name is making it unclear.
0b0000 = fwd based on inner for all tagged,untagged and double-tagged packets
0b1111 = fwd based on outer for all tagged,untagged and double-tagged packets

see, https://gitlab.com/olliver/openwrt/realtek_sdk/-/blob/xgs1250/sdk/src/dal/longan/dal_longan_vlan.c?ref_type=heads#L3037
and
https://gitlab.com/olliver/openwrt/realtek_sdk/-/blob/xgs1250/sdk/include/common/rt_type.h?ref_type=heads#L234

0xharshal replied:
New change reverts back the original code in dsa.c and changes rtl93xx vlan_fwd_on_inner functions to match rtl83xx vlan_fwd_on_inner, 

I have verified it with rtl930x.
I want to verify QinQ behavior on rtl83xx and rtl931x - as of now the code looks correct, but if in future someone complains about QinQ, we know where to look

simonwunderlich commented:
> Please also clarify what Simon's involvement is in these patches, or I'm going to have to ask you to drop his Signed-off-by.

@svanheule I've been involved in the development and debugging, although the vast majority of the actual work was done by Harshal. I've also reviewed the patches and we had some internal iterations on this patchset.

If you prefer, we can add my acked-by or reviewed-by if you feel that is more appropriate.

0xharshal commented on target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.h:
```diff
@@ -597,7 +597,7 @@ typedef enum {
 #define MAX_HOST_ROUTES 1536
 #define MAX_INTF_MTUS 8
-#define DEFAULT_MTU 1536
+#define DEFAULT_MTU 9000
```
Dropped this change

0xharshal commented on target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.h:
```diff
@@ -216,6 +216,9 @@
 
+/* Default MTU with jumbo frames support */
+#define DEFAULT_MTU 9000
+
```
kept this one

0xharshal commented on target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c:
```diff
@@ -2396,23 +2396,61 @@ static void rtl930x_led_init(struct rtl838x_switch_priv *priv)
+			break;
+		}
+
+		if (leds_in_this_set == 0 || leds_in_this_set > sizeof(set_config)) {
```
yes, good spot. fixed in latest push.

svanheule commented on target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c:
```diff
@@ -2396,23 +2396,62 @@ static void rtl930x_led_init(struct rtl838x_switch_priv *priv)
+#define RTL930X_LED_SETX_LEDY(x,y) RTL930X_LED_SETX_0_CTRL(x) - 4*(y/2)
+
+// get shift for given led in any set
+#define RTL930X_LED_SET_LEDX_SHIFT(x) (16*(x%2))
```
This gives the impression that these definitions are scoped, when they aren't. Just place them at the top of the file with the other `#define`s. It would also be prudent to enclose all the definitions in parentheses, like you did for `RTL930X_LED_SET_LEDX_SHIFT(x)`. And as a final nit, kernel style is to add spaces around operators: `(16 * (x % 2))`

0xharshal replied:
done

svanheule commented:
Thanks for the updates and for trimming things down to just the required changes! Sorry I didn't make that remark on the `#define`s yesterday.

> @svanheule I've been involved in the development and debugging, although the vast majority of the actual work was done by Harshal. I've also reviewed the patches and we had some internal iterations on this patchset.
> 
> If you prefer, we can add my acked-by or reviewed-by if you feel that is more appropriate.

Thanks for the info, @simonwunderlich! From the sound of it, Co-developed-by (with Signed-off-by) would've been appropriate too. If you two are OK with Acked-by, we can give you credit that way too. We'll still know where to find you ;-)

@TobleMiner would you like to give these patches a spin before they are merged? Feel free to leave a Tested-by.

Borromini commented:
To be sure this is not a PEBKAC issue - is network connectivity supposed to work with the default 'boota' command? Because after loading a build from main HEAD with your last revision, the switch seems unreachable over the network if I revert to the default 'boota'.

Does one need to stick with the 'rtk network on' command?

`fw_printenv` output:

```
SN=S202Lxxxxxxxx
baudrate=115200
boardmodel=XGS1250_12
bootdelay=1
ethact=rtl9300#0
ethaddr=BC:CF:xx:xx:xx:xx
fan_high=12000
fan_low=1500
fan_start=1
ipaddr=192.168.1.1
ledModeInitSkip=0
serverip=192.168.1.111
stderr=serial
stdin=serial
stdout=serial
tmp_high=108
bootcmd=boota
```

Thanks

Borromini commented:
I'll assume one still needs `rtk network on`, with that set, I tested 2,5 Gbps on a RTL8125 client.

Serial prints the following when port 9 gets connected, looks good:

```
Sat Jan  6 10:56:53 2024 kern.info kernel: [  194.821335] rtl93xx_phylink_mac_config port 24, mode 0, phy-mode: usxgmii, speed 2500, link 1
Sat Jan  6 10:56:53 2024 kern.info kernel: [  194.830892] rtl93xx_phylink_mac_config SDS is 6
Sat Jan  6 10:56:53 2024 kern.info kernel: [  194.835995] rtl83xx-switch switch@1b000000 lan9: Link is Up - 2.5Gbps/Full - flow control rx/tx
Sat Jan  6 10:56:53 2024 kern.info kernel: [  194.846558] switch: port 9(lan9) entered blocking state
Sat Jan  6 10:56:53 2024 kern.info kernel: [  194.852465] switch: port 9(lan9) entered forwarding state
Sat Jan  6 10:56:53 2024 daemon.notice netifd: Network device 'lan9' link is up
Sat Jan  6 10:56:53 2024 kern.info kernel: [  194.860804] rtl93xx_phylink_mac_config port 24, mode 0, phy-mode: usxgmii, speed 2500, link 1
Sat Jan  6 10:56:53 2024 kern.info kernel: [  194.870385] rtl93xx_phylink_mac_config SDS is 6
Sat Jan  6 10:57:14 2024 daemon.notice netifd: Network device 'lan9' link is down
Sat Jan  6 10:57:14 2024 kern.info kernel: [  216.656203] rtl83xx-switch switch@1b000000 lan9: Link is Down
Sat Jan  6 10:57:14 2024 kern.info kernel: [  216.662700] switch: port 9(lan9) entered disabled state
Sat Jan  6 10:57:19 2024 kern.info kernel: [  220.821482] rtl93xx_phylink_mac_config port 24, mode 0, phy-mode: usxgmii, speed 2500, link 1
Sat Jan  6 10:57:19 2024 kern.info kernel: [  220.831036] rtl93xx_phylink_mac_config SDS is 6
Sat Jan  6 10:57:19 2024 kern.info kernel: [  220.836139] rtl83xx-switch switch@1b000000 lan9: Link is Up - 2.5Gbps/Full - flow control rx/tx
Sat Jan  6 10:57:19 2024 kern.info kernel: [  220.846709] switch: port 9(lan9) entered blocking state
Sat Jan  6 10:57:19 2024 kern.info kernel: [  220.852606] switch: port 9(lan9) entered forwarding state
Sat Jan  6 10:57:19 2024 daemon.notice netifd: Network device 'lan9' link is up
```

Ethtool output looks good as well:

```
$ sudo ethtool eth0
Settings for eth0:
        Supported ports: [ TP    MII ]
        Supported link modes:   10baseT/Half 10baseT/Full
                                100baseT/Half 100baseT/Full
                                1000baseT/Full
                                2500baseT/Full
        Supported pause frame use: Symmetric Receive-only
        Supports auto-negotiation: Yes
        Supported FEC modes: Not reported
        Advertised link modes:  10baseT/Half 10baseT/Full
                                100baseT/Half 100baseT/Full
                                1000baseT/Full
                                2500baseT/Full
        Advertised pause frame use: Symmetric Receive-only
        Advertised auto-negotiation: Yes
        Advertised FEC modes: Not reported
        Link partner advertised link modes:  100baseT/Half 100baseT/Full
                                             1000baseT/Full
                                             10000baseT/Full
                                             2500baseT/Full
                                             5000baseT/Full
        Link partner advertised pause frame use: Symmetric Receive-only
        Link partner advertised auto-negotiation: Yes
        Link partner advertised FEC modes: Not reported
        Speed: 2500Mb/s
        Duplex: Full
        Auto-negotiation: on
        master-slave cfg: preferred slave
        master-slave status: slave
        Port: Twisted Pair
        PHYAD: 0
        Transceiver: external
        MDI-X: Unknown
        Supports Wake-on: pumbg
        Wake-on: d
        Link detected: yes
```

I have no other 2,5 Gbps client so I am unable to test actual throughput, but I'm able to connect over the link. The triangular LED for port 9 lights up blue as well.

svanheule commented:
Then without further ado, merged to main. Thanks everyone!

Borromini commented:
@0xharshal When trying to port the XGS1210-12 rev B1, I noticed this error in dmesg. Checked the XGS1250-12 with a build from today's main tree, and it's showing up there as well. Any idea what this is pointing to? I haven't altered anything on the XGS1250-12.

`[    2.104605] rtl930x_led_init led_set configuration invalid skipping over this set`

0xharshal commented:
Hello @Borromini check the dts file https://github.com/0xharshal/openwrt/blob/917156592aa0842a7247b0b8a764f88c306613cd/target/linux/realtek/dts-5.15/rtl9302_zyxel_xgs1250-12.dts#L65, 

this patch changes the way led_set{0,1,2,3} is interpreted

Borromini commented:
I'm aware yes, your patch modifies the XGS1250-12 DTS. I assume this is expected behaviour then?

0xharshal commented:
Hmm, I don't think this is expected, it should only print this error when led_set is misconfigured, I need some time to investigate on my hardware (which is not xgs125-12 but based on rtl9302)

TobleMiner commented:
Can confirm this error in dmesg.

I think `of_property_count_u32_elems` [can return negative error values](https://elixir.bootlin.com/linux/v5.15.148/source/include/linux/of.h#L980). The code [here](https://github.com/openwrt/openwrt/blob/1c02b874fa795dbfdc008d98cd58b59613b923c7/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c#L2424C3-L2424C19) does not seem to handle that case.
I think comparison for `leds_in_this_set > sizeof(set_config)` will catch most negative values due to signed to unsigned conversion but that line really should be 
```if (leds_in_this_set <= 0 || leds_in_this_set > sizeof(set_config)) {```

In general it seems like `rtl930x_led_init` assume that all led sets (0-3) are always present and prints a warning about invalid configuration if one is not (`of_property_count_u32_elems` returns `-EINVAL`  in that case).  
In contrast `rtl83xx_mdio_probe` will ignore a missing led set happily though, [storing `-EINVAL` in `leds_on_this_port`](https://github.com/openwrt/openwrt/blob/1c02b874fa795dbfdc008d98cd58b59613b923c7/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/common.c#L367)

Since it seems that led sets are not mandatory and need to be referenced explicitly by their ports I think it would make sense to ignore missing LED sets entirely. Or if we want to be fancy we could validate if all sets referenced by ports exist. Maybe [`of_property_present`](https://elixir.bootlin.com/linux/v5.15.148/C/ident/of_property_present) could be useful here?
